File: demos/solar-system/reference/SolarSystem.java
import java.awt.*;
public class SolarSystem
{
    private SimpleGrid env;
    private Particle sun;
    private final int x = 230,y=300;

    
    public SolarSystem()
    {
        env = new SimpleGrid();
        sun = new Particle(new Point(x,y),45);
        
        //balance frame of reference
        //sun.myVelocity.add(new Vector(0,-.017));
        
        sun.setDensity(50);
        env.add(sun);
        init();

        
    }
    public SimpleGrid getEnv()
    { return env; }
    private void init()    
    {
        for (int i=2;i<=7;i++)//3,5
        {
            Particle p = new Particle(new Point(x+150*i,y),7);
            p.setDensity(13);
            addVelocity(sun,p);
            
            //balance momentum in order to get a stationary frame of reference
            double mv = p.myVelocity.magnitude()*p.getMass();
            sun.myVelocity.add(new Vector(0,-1).dilate(mv/sun.getMass()));
            
            env.add(p);
            if (i>3)//3
            {
                Particle r = new Particle(new Point(p.center().x,p.center().y-12),2);
                r.setDensity(.01);
                r.myVelocity=(Vector)p.myVelocity.clone();
                addVelocity(p,r);
                env.add(r);
            }
        }
    }
    
    //sets velocity to a stable orbit
    private void addVelocity(Particle big, Particle orbit)
    {
        double velocity = Math.sqrt((2*Particle.G*big.getMass())/distanceBetween(big.center(),orbit.center()));
        Vector v = new Vector(big.center(),orbit.center());
        v.rotate(Math.PI/2.0);
        orbit.myVelocity.add(v.setMagnitude(.6*velocity));
    }
    
    private double distanceBetween(Point a, Point b)
    { return Math.sqrt(Math.pow(a.x-b.x,2)+Math.pow(a.y-b.y,2)); }
    
    //public void show(java.awt.Graphics g)
    //{ env.show(g); }
    
    public void step(Graphics g)
    {
        //sun.moveTo(new Point(x,y));
        env.step(g);
    }
    
}

File: demos/solar-system/solar-system.js
class SolarSystem
{

    constructor()
    {
        this.x = 230;
        this.y = 300;

        this.env = new SimpleGrid();
        this.sun = new Particle(new Point(this.x,this.y),45);

        //balance frame of reference
        //sun.myVelocity.add(new Vector(0,-.017));

        this.sun.setDensity(50);
        this.env.add(this.sun);
        this.init();
    }

    getEnv()
    {
        return this.env;
    }

    init()
    {
        for (let i=2;i<=7;i++)//3,5
        {
            let p = new Particle(new Point(this.x+150*i,this.y),7);
            p.setDensity(13);
            this.addVelocity(this.sun,p);

            //balance momentum in order to get a stationary frame of reference
            let mv = p.myVelocity.magnitude()*p.getMass();
            this.sun.myVelocity.add(Vector.fromDimensions(0,-1).dilate(mv/this.sun.getMass()));

            this.env.add(p);
            if (i>3)//3
            {
                let r = new Particle(new Point(p.center().x,p.center().y-12),2);
                r.setDensity(.01);
                r.myVelocity=p.myVelocity.clone();
                this.addVelocity(p,r);
                this.env.add(r);
            }
        }
    }

    //sets velocity to a stable orbit
    // big - particle
    // orbit - particle
    addVelocity(big, orbit)
    {
        let velocity = Math.sqrt((2*Particle.G()*big.getMass())/this.distanceBetween(big.center(),orbit.center()));
        let v = Vector.fromEndPoints(big.center(),orbit.center());
        v.rotate(Math.PI/2.0);
        orbit.myVelocity.add(v.setMagnitude(.6*velocity));
    }

    // a - Point
    // b - Point
    // returns double
    distanceBetween(a, b)
    { return Math.sqrt(Math.pow(a.x-b.x,2)+Math.pow(a.y-b.y,2)); }

    step(ctx)
    {
        this.env.step(ctx);
    }

}